nestkernel/node_collection.cpp

```diff
@@ -283,14 +333,21 @@ NodeCollectionPrimitive::NodeCollectionPrimitive( index first,
   assert( first_ <= last_ );
 }
 
-NodeCollectionPrimitive::NodeCollectionPrimitive( index first, index last, index model_id )
+NodeCollectionPrimitive::NodeCollectionPrimitive( index first, index last, index model_id, bool for_testing )
```

Comment:
I am not really happy with the flag here, and actually the fact that you need the additional create function for a mock version of `NodeCollectionPrimitive`. If I understand the code below correctly, you only need the distinction because the `get_node_model()` call below will fail in C++ test cases not including the full kernel. 

To build the code below, it clearly needs to be linked against the full kernel. I see several possibilities:

1. Add `get_num_models()` to `ModelManager` and execute the `nodes_have_no_proxies_` line only if there are models, otherwise, set to `false`. Could even be a 
```c++
nodes_have_no_proxies_ = kernel().model_manager.get_num_models() > 0 and not kernel().model_manager.get_node_model( model_id_ )->has_proxies();
```
2. Put the call in a try-except and just ignore errors. This is dangerous if something should really be wrong. It could be wrapped using a compiler flag such as `NEST_CXX_TEST` and re-throw if this is off, otherwise pass silently.
3. Create a class `NodeCollectionPrimitiveBase` which contains all of `NodeCollectionPrimitive` but fixes `nodes_have_no_proxies_` to `false`. Then derive `NodeCollectionPrimitive` from it, and set the field correctly. The, make the `CollectionT` a template argument of `create_test_collection()` with `NodeCollectionPrimitive` as default value. In the test program itself, you can then derive `NodeCollectionPrimitiveMocker` from `NodeCollectionPrimitiveBase` and use that as template argument for `create_test_collection()`; you could probably also use `...Base` directly.  With that in place, you can then use the `create_test_collection()` as a general mechanism to build a node collection from arbitrary arrays of GIDs.

I think #3 is the most general and powerful approach which also keeps all test-specific stuff in the test.

Reply:
No. 3 Would definitely be the way to go if we continue on adding more C++ tests (incl. mocking) in the future. From a design perspective, this would be very clean.
However, if we don't plan on adding any more mocks in the near future, a simple compiler flag wrapping test code would be sufficient as well in my opinion.

Reply:
I expect that we will have more C++ tests in the future. We therefore should start out with a good design as we all have a tendency to build new code on existing examples. Good examples at the beginning are therefore crucial.

One thing I am not entirely happy with in 3. is the need to introduce the additional `...Base` class just to make mocking possible. I experimented a bit to see if one could have the dependency on `kernel()` in a base class and the define it away in a derived class (so that the ...`Mocker` class could derive directly from `NodeCollectionPrimitive`), but since class base constructors never call derived-class methods, I haven't found a way to achieve that.

Reply:
I might have a different idea, but I don't know how clean it would sound. 

We can use a **forward-declaration** of a class. This class will be called `NodeCollectionInitializer` (for each target class to test, we create a` [TargetClassName]Initializer`).  In the `TargetClass` (i.e, `NodeCollectionPrimitive`), we add a new constructor that takes a `NodeCollectionPrimitiveInitializer` as an argument.

This `NodeCollectionPrimitiveInitializer` is just a simple class that stores information about the `TargetClass` and provides only a `constructor` and `Getter` functions. 

The new added `constructor` in the `TargetClass` will not be defined in the `TargetClass`'s declaration nor definition inside the nestkernel, but instead the new  `constructor`  will be defined inside the cppunittest folder. So when compiling nest, the  `constructor` won't be used, and also we will not have a linker problem, but when compiling the unitests, we add the file that has the definition of the new  `constructor`.

In the definition of the new  `constructor` in the `TagetClass`, we simply use the ` [TargetClassName]Initializer` to set to  the values that we want, and therefore we can bypass those provided constructors that use the `kernel` or do any other unwanted setups.

Reply:
Given that each `TargetClass` will only have one mock class (only one `[TargetClassName]Initializer`), thus we will not pollute the header file of the  `TargetClass`  with many **forward-declarared** mocks. 

The `TargetClass` will just use the  `[TargetClassName]Initializer` to initiate its inner state, and we simply can use the instances of the `TargetClass` normally without any restrictions.

Reply:
## Example
in `nestkernel/node_collection.h`
```
namespace nest {
class NodeCollectionPrimitiveInitializer;
class NodeCollectionPrimitive 
  {
       public:
       /....../
          NodeCollectionPrimitive(const NodeCollectionPrimitiveInitializer& mock);
  };
}
```
in `cpptests/node_collection_mock.h`
```
#ifndef TEST_H
#define TEST_H
namespace nest {
     class NodeCollectionPrimitiveInitializer 
     {
           //store the needed information here
     };
}
nest::NodeCollectionPrimitive:: NodeCollectionPrimitive(const NodeCollectionPrimitiveInitializer& mock)
{
       //retreive information from mock object.
       first = mock.get_first();
       last = mock.get_last();

}
#endif
```

Reply:
> One thing I am not entirely happy with in 3. is the need to introduce the additional `...Base` class just to make mocking possible. I experimented a bit to see if one could have the dependency on `kernel()` in a base class and the define it away in a derived class (so that the ...`Mocker` class could derive directly from `NodeCollectionPrimitive`), but since class base constructors never call derived-class methods, I haven't found a way to achieve that.

I think this is the cleanest solution. We don't have to call derived-class methods if we bypass the `create` functions and directly create the NodeCollection in the test.